File: web/src/val/report.ts
// The capability report, derived from the code rather than declared by its
// author (spec §7). Every line here is answerable statically only because of
// decisions taken for other reasons — effects live in `execute`, strings cannot
// be assembled, audiences are fixed in the manifest, the language is total.

import type { Decl, Node, Program } from './parse'

export type ReportLine = { label: string; values: string[]; tone?: 'plain' | 'warn' }
export type Finding = { line: number; message: string; severity: 'error' | 'warning' }

export type Report = {
  app: string
  version: string
  lines: ReportLine[]
  findings: Finding[]
}

const flatten = (nodes: Node[]): Node[] => nodes.flatMap((n) => [n, ...flatten(n.children)])

export function report(program: Program): Report {
  const { decls } = program
  const app = (decls.find((d) => d.t === 'app') as any)?.id ?? '—'
  const version = (decls.find((d) => d.t === 'version') as any)?.value ?? '—'
  const caps = (decls.find((d) => d.t === 'capabilities') as any)?.entries ?? ([] as Node[])
  const actions = decls.filter((d) => d.t === 'action') as Extract<Decl, { t: 'action' }>[]
  const screens = decls.filter((d) => d.t === 'screen') as Extract<Decl, { t: 'screen' }>[]
  const trusts = decls.filter((d) => d.t === 'trust') as Extract<Decl, { t: 'trust' }>[]

  const effects = actions.flatMap((a) => flatten(Object.values(a.phases).flat()))
  const findingsEarly: Finding[] = []
  const findings: Finding[] = []
  const used = new Set<string>()

  // reads --------------------------------------------------------------------
  const reads = new Set<string>()
  for (const s of screens)
    for (const d of s.data)
      if (d.source === 'credentials' && d.credentialType)
        reads.add(`${d.credentialType}${d.policy ? ` under ${d.policy}` : ' — unverified'}`)
  for (const a of actions) {
    const v = a.raw['verify'] ?? ''
    for (const t of trusts) if (v.includes(t.name)) reads.add(`${t.subjectType || '?'} under ${t.name}`)
    const inp = a.raw['input'] ?? ''
    const m = inp.match(/Credential<\s*(\w+)/)
    if (m && ![...reads].some((r) => r.startsWith(m[1]))) reads.add(`${m[1]} — unverified`)
  }
  if (reads.size) used.add('credential.read')

  // discloses / proves -------------------------------------------------------
  // Counted per action, because the rule is per action: a second disclosure
  // cannot be conditional on a batch the first has already completed (spec §5).
  const discloses: string[] = []
  const proves: string[] = []
  for (const a of actions) {
    let here = 0
    for (const line of (a.raw['execute'] ?? '').split('\n')) {
      const d = line.match(/\bdisclose\s+(.+)/)
      if (d) { discloses.push(d[1].trim()); here++ }
      const pr = line.match(/\bprove\s+(.+)/)
      if (pr) { proves.push(pr[1].trim()); here++ }
    }
    if (here > 1)
      findingsEarly.push({
        line: a.line,
        severity: 'error',
        message: `\`${a.name}\` performs ${here} disclosures. An action performs at most one: a second cannot be conditional on a batch the first has already completed.`,
      })
  }
  if (discloses.length || proves.length) used.add('disclosure.present')

  // issues -------------------------------------------------------------------
  const issues = new Set<string>()
  for (const n of effects) if (n.kind === 'credential.issue') issues.add(n.args[0]?.value.split(/[\s{]/)[0] ?? '?')
  if (issues.size) used.add('credential.issue')

  // audiences ----------------------------------------------------------------
  const audiences = new Set<string>()
  for (const c of caps as Node[]) {
    if (c.kind === 'api.query') {
      const a = c.args.find((x) => x.name === 'audience')
      if (a) audiences.add(a.value.replace(/"/g, ''))
    }
  }
  for (const s of screens) for (const d of s.data) if (d.source === 'query' && d.audience) audiences.add(d.audience)
  if (audiences.size) used.add('api.query')

  // payments -----------------------------------------------------------------
  const payments: string[] = []
  for (const n of effects) if (n.kind === 'payment.request') payments.push(n.args.map((a) => `${a.name ? a.name + ': ' : ''}${a.value}`).join(', '))
  if (payments.length) used.add('payment.request')

  // state --------------------------------------------------------------------
  const writes = new Set<string>()
  for (const a of actions)
    for (const line of (a.raw['update'] ?? '').split('\n')) {
      const m = line.match(/^\s*([\w.]+)\s*:/)
      if (m) writes.add(m[1])
    }

  // findings -----------------------------------------------------------------
  for (const c of caps as Node[]) {
    const head = c.kind.split('(')[0]
    if (!used.has(head))
      findings.push({
        line: c.line,
        severity: 'error',
        message: `\`${c.kind}\` is declared and never used. Consent asked for something unused is consent spent on nothing, and it trains people to say yes.`,
      })
  }
  findings.push(...findingsEarly)
  for (const s of screens)
    for (const d of s.data)
      if (d.source === 'credentials' && !d.policy)
        findings.push({ line: d.line, severity: 'warning', message: `\`${d.name}\` is read without \`verified with\`, so the list can show a credential nothing vouched for.` })

  const lines: ReportLine[] = [
    { label: 'reads', values: [...reads] },
    { label: 'discloses', values: discloses.length ? discloses : ['nothing'] },
    { label: 'proves', values: proves },
    { label: 'issues', values: [...issues] },
    { label: 'talks to', values: [...audiences] },
    { label: 'moves money', values: payments, tone: 'warn' },
    { label: 'writes state', values: [...writes] },
    {
      label: 'irreversible',
      values: discloses.length || proves.length ? ['one disclosure'] : ['none'],
      tone: discloses.length ? 'warn' : 'plain',
    },
  ]

  return { app, version, lines, findings }
}

File: crates/valang/src/report.rs
//! The capability report (§7): what this application can do to the person,
//! derived from the code rather than written by its author.
//!
//! A publisher ships a copy for review and it is evidence of nothing — the host
//! derives this one and refuses on mismatch.
//!
//! **What it does to the person is not filled in here.** That comes from the
//! compiled module, whose import section is the whole of what it can reach —
//! see `valang_wasm::report_of`, which is what a caller wants. This half is
//! what the front end is the authority on: which application it is, which hosts
//! it needs, what it exports and what it takes from other packages.
//!
//! There were two routes for a while and they disagreed the first time anybody
//! compared them: the walk over the source said a program proving an age read
//! the birthdate, about a module that has no way to reach it. One route, and
//! this is the half that stayed.

use std::collections::BTreeSet;
use std::fmt;

use crate::ast::*;

#[derive(Debug, Default, PartialEq, Eq)]
pub struct Report {
    pub app: String,
    pub version: String,
    pub reads: BTreeSet<String>,
    /// Credentials checked against a policy and **not read**. A person is owed
    /// the difference: an application that checks your national ID against the
    /// government's key and reads nothing off it is not an application that
    /// reads your national ID.
    pub checks: BTreeSet<String>,
    pub discloses: BTreeSet<String>,
    pub proves: BTreeSet<String>,
    pub issues: BTreeSet<String>,
    pub audiences: BTreeSet<String>,
    pub payments: BTreeSet<String>,
    pub writes: BTreeSet<String>,
    /// The hosts this package needs beyond the registry every one of them
    /// provides. Empty means it runs on any of them, which is a sentence a
    /// person deciding to install it is owed as much as the rest.
    pub hosts: BTreeSet<String>,
    /// The addresses a link from outside can reach. A package that can be
    /// opened by a link is a package whose screens somebody else can point at,
    /// which is a sentence the person installing it is owed.
    pub addresses: BTreeSet<String>,
    /// What this package exports, with the parameters each one takes.
    ///
    /// The surface somebody else's build depends on. Changing a parameter here
    /// breaks a package that is not yours and whose author is not in the room,
    /// so it belongs in the report rather than only in the source: a publisher
    /// can see what they are about to change, and diff it against what they
    /// published.
    pub exports: BTreeSet<String>,
    /// What this package takes from others, as written.
    pub imports: BTreeSet<String>,
    pub irreversible: bool,
}

pub fn report(p: &Program) -> Report {
    let mut r = Report {
        app: p.app.clone().unwrap_or_default(),
        version: p.version.clone().unwrap_or_default(),
        ..Default::default()
    };

    r.hosts = p
        .hosts
        .iter()
        .filter(|h| !h.starts_with(crate::capability::CORE))
        .cloned()
        .collect();
    r.addresses = p
        .screens
        .iter()
        .flat_map(|s| &s.settings)
        .filter(|a| a.name.as_deref() == Some("address"))
        .filter_map(|a| match &a.value {
            Expr::Str { value, .. } => Some(value.clone()),
            _ => None,
        })
        .collect();
    r.exports = p
        .components
        .iter()
        .filter(|c| c.exported)
        .map(|c| {
            let params: Vec<String> = c
                .params
                .iter()
                .map(|f| format!("{}: {}", f.name, f.ty.written()))
                .collect();
            format!("{}({})", c.name, params.join(", "))
        })
        .collect();

    r.imports = p
        .imports
        .iter()
        .map(|i| format!("{} {{ {} }}", i.package, i.names.join(", ")))
        .collect();

    r.irreversible = !r.discloses.is_empty() || !r.proves.is_empty() || !r.payments.is_empty();
    r
}

/// Every statement, including the ones inside a branch.
///
/// A flat loop over a phase's statements is a report that stops at the first
/// `if`: an effect written in one branch never appeared, so the consent sheet —
/// which is a rendering of this report — did not mention something the
/// application does.
/// `checked.claims.country` is what the author wrote; `NationalId.country` is
/// what the person is asked about, and every sheet is read by the second one.
///
/// **One place.** The compiler names an import with it, the evaluator names a
/// statement it hands to the host with it, and the report renders with it — and
/// when there were two, a proof came out named after a local binding in one and
/// after the credential in the other, which is two different statements in two
/// records of the same run.
pub fn in_credential_terms(p: &Program, said: &str) -> String {
    let mut bound = Vec::new();
    for a in &p.actions {
        for block in &a.phases {
            verified_in(&block.stmts, &mut bound);
        }
    }
    let mut out = said.to_string();
    for (name, policy) in bound {
        let Some(t) = p.trusts.iter().find(|t| t.name == policy) else { continue };
        out = out.replace(&format!("{name}.claims."), &format!("{}.", t.subject_type));
    }
    out
}

/// Every `x with Policy` in a body, including the ones inside a branch — a
/// binding made in one names a credential the same way.
fn verified_in(stmts: &[Stmt], out: &mut Vec<(String, String)>) {
    for s in stmts {
        match s {
            Stmt::Let { name, value: Expr::With { policy, .. }, .. } => {
                out.push((name.clone(), policy.clone()))
            }
            Stmt::If { then, other, .. } => {
                verified_in(then, out);
                verified_in(other, out);
            }
            Stmt::Effect { body, .. } => verified_in(body, out),
            _ => {}
        }
    }
}

/// How a predicate or an argument is written down where a person reads it.
///
/// Public because the back end names an import with it — `prove:` carries the
/// statement it proves, and the two have to be the same string or the report a
/// wallet derives from a module would not be the report this says.
pub fn render(e: Option<&Expr>) -> String {
    match e {
        None => "—".into(),
        Some(Expr::Num { value, .. }) => value.to_string(),
        Some(Expr::Str { value, .. }) => format!("\"{value}\""),
        Some(Expr::Binary { op, lhs, rhs, .. }) => format!("{} {op} {}", render(Some(lhs)), render(Some(rhs))),
        Some(Expr::Call { callee, .. }) => format!("{}(…)", callee.path().unwrap_or_else(|| "?".into())),
        Some(other) => other.path().unwrap_or_else(|| "…".into()),
    }
}

impl fmt::Display for Report {
    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
        let row = |f: &mut fmt::Formatter<'_>, label: &str, vals: &BTreeSet<String>| -> fmt::Result {
            if vals.is_empty() {
                writeln!(f, "{label:<14} —")
            } else {
                for (i, v) in vals.iter().enumerate() {
                    if i == 0 {
                        writeln!(f, "{label:<14} {v}")?;
                    } else {
                        writeln!(f, "{:<14} {v}", "")?;
                    }
                }
                Ok(())
            }
        };
        writeln!(f, "{} v{}", self.app, self.version)?;
        row(f, "reads", &self.reads)?;
        row(f, "checks", &self.checks)?;
        row(f, "discloses", &self.discloses)?;
        row(f, "proves", &self.proves)?;
        row(f, "issues", &self.issues)?;
        row(f, "talks to", &self.audiences)?;
        row(f, "moves money", &self.payments)?;
        row(f, "writes state", &self.writes)?;
        row(f, "runs only on", &self.hosts)?;
        row(f, "reachable at", &self.addresses)?;
        row(f, "exports", &self.exports)?;
        row(f, "imports", &self.imports)?;
        writeln!(f, "{:<14} {}", "irreversible", if self.irreversible { "yes" } else { "none" })
    }
}
